indico/modules/rb/client/js/common/bookings/BookingDetails.jsx

```diff
@@ -114,6 +118,76 @@ class BookingDetails extends React.Component {
     );
   };
 
+  renderCancelOccurrence = (bookings, id) => {
+    const {
+      actions: {openBookingDetails},
+    } = this.props;
+    const {cancelDate} = this.props;
+    const serializedDate = serializeDate(cancelDate, 'L');
+    if (cancelDate === null) {
+      return null;
+    } else if (bookings[cancelDate] === undefined || bookings[cancelDate].length === 0) {
+      return (
+        <Modal open size="tiny" onClose={() => openBookingDetails(id)}>
+          <Modal.Header>Cancellation not possible</Modal.Header>
+          <Modal.Content>
+            <Translate>
+              There is no occurrence on this date (
+              <Param name="serializedDate" value={serializedDate} />
+              ).
+            </Translate>
+          </Modal.Content>
+          <Modal.Actions>
+            <Button content={Translate.string('Close')} onClick={() => openBookingDetails(id)} />
+          </Modal.Actions>
+        </Modal>
+      );
+    } else if (!bookings[cancelDate][0].canCancel) {
+      return (
+        <Modal open size="tiny" onClose={() => openBookingDetails(id)}>
+          <Modal.Header>
+            <Translate>Cancellation not possible</Translate>
+          </Modal.Header>
+          <Modal.Content>
+            <Translate>
+              The occurrence you chose (
+              <Param name="serializedDate" value={serializedDate} />) cannot be cancelled.
+            </Translate>
+          </Modal.Content>
+          <Modal.Actions>
+            <Button content={Translate.string('Close')} onClick={() => openBookingDetails(id)} />
+          </Modal.Actions>
+        </Modal>
+      );
+    } else {
+      return (
+        <Modal open size="tiny" onClose={() => openBookingDetails(id)}>
+          <Modal.Header>
+            <Translate>Confirm cancellation</Translate>
+          </Modal.Header>
+          <Modal.Content>
+            <Translate>
+              Are you sure you want to cancel this occurrence (
+              <Param name="serializedDate" value={serializedDate} />
+              )?
+            </Translate>
+          </Modal.Content>
+          <Modal.Actions>
+            <Button content={Translate.string('Close')} onClick={() => openBookingDetails(id)} />
+            <Button
+              content={Translate.string('Cancel this occurrence')}
+              onClick={() => {
+                this.changeOccurrenceState('cancel');
+                openBookingDetails(id);
+              }}
```

Comment:
It would be safer to make this function async and await the `this.changeOccurrenceState` call to make sure the details don't open until after you fully finished cancelling